src/test/groovy/graphql/execution/instrumentation/TestBatchLoadingSupported.groovy

```diff
@@ -0,0 +1,206 @@
+                ])
+        )
+                .type(newTypeWiring("Human")
+                .dataFetcher("friends", friendsDataFetcher)
+        )
+                .type(newTypeWiring("Droid")
+                .dataFetcher("friends", friendsDataFetcher)
+        )
+
+                .type(newTypeWiring("Character")
+                .typeResolver(characterTypeResolver)
+        )
+                .type(episodeWiring)
+                .build()
+        wiring
+    }
+
+    def schema = schemaFile("starWarsSchema.graphqls", starWarsWiring())
+
+
+    def "basic batch loading is possible via instrumentation interception of Execution Strategies"() {
+
+        given:
+        def batchingInstrumentation = new NoOpInstrumentation() {
+            @Override
+            InstrumentationContext<CompletableFuture<ExecutionResult>> beginExecutionStrategy(InstrumentationExecutionStrategyParameters parameters) {
+                return new InstrumentationContext<CompletableFuture<ExecutionResult>>() {
+                    @Override
+                    void onEnd(CompletableFuture<ExecutionResult> result, Throwable t) {
+                        //
+                        // this causes "batched" futures to actually be turned into batch loads
+                        characterDataLoader.dispatch()
+                    }
+                }
+            }
+        }
+
+        def graphql = GraphQL.newGraphQL(schema).instrumentation(batchingInstrumentation).build()
+
+        when:
+        def query = """
+        query {
+            hero {
+                name 
+                friends {
+                    name
+                    friends {
+                       name
+                    } 
+                }
+            }
+        }
+        """
+
+        def asyncResult = graphql.executeAsync(ExecutionInput.newExecutionInput().query(query))
+
+        def er = asyncResult.join()
+
+        then:
+        er.data == [hero: [name: 'R2-D2', friends: [
+                [name: 'Luke Skywalker', friends: [
+                        [name: 'Han Solo'], [name: 'Leia Organa'], [name: 'C-3PO'], [name: 'R2-D2']]],
+                [name: 'Han Solo', friends: [
+                        [name: 'Luke Skywalker'], [name: 'Leia Organa'], [name: 'R2-D2']]],
+                [name: 'Leia Organa', friends: [
+                        [name: 'Luke Skywalker'], [name: 'Han Solo'], [name: 'C-3PO'], [name: 'R2-D2']]]]]
+        ]
+
+        //
+        // there are five characters in this query however we have asked for their details over and over
+        // and yet we only actually load the objects up five times
+        rawCharacterLoadCount == 5
+        //
+        // our batch load API only gets called thrice, for R2D2, for his friends and for their friends
+        batchFunctionLoadCount == 3
+        //
+        // if we didn't have batch loading it would have these many character load calls
+        naiveLoadCount == 15
+    }
+}
```

Comment:
These numbers are startling.  In  a naive world where you need to make a HTTP call to get "character" data you would end up with 15 HTTP calls.

Here we have at most 5 calls and if you have a HTTP API that accepts multiple 'character ids' then we only make 3 calls to it.